ert/data/record/_record.py

```diff
@@ -0,0 +1,224 @@
+import pathlib
+from abc import ABC
+from collections import deque
+from enum import Enum
+from typing import Dict, List, Mapping, MutableMapping, Optional, Tuple, Union, cast
+
+from beartype import beartype
+from beartype.roar import BeartypeException  # type: ignore
+from pydantic import PositiveInt
+from ert.serialization import get_serializer
+
+number = Union[int, float]
+numerical_record_data = Union[
+    List[number],
+    Dict[str, number],
+    Dict[int, number],
+]
+blob_record_data = bytes
+record_data = Union[numerical_record_data, blob_record_data]
+record_collection = Tuple["Record", ...]
+RecordIndex = Tuple[Union[int, str], ...]
+
+
+def _build_record_index(
+    data: numerical_record_data,
+) -> RecordIndex:
+    if isinstance(data, MutableMapping):
+        return tuple(data.keys())
+    else:
+        return tuple(range(len(data)))
+
+
+class RecordValidationError(Exception):
+    pass
+
+
+class RecordType(str, Enum):
+    LIST_FLOAT = "LIST_FLOAT"
+    MAPPING_INT_FLOAT = "MAPPING_INT_FLOAT"
+    MAPPING_STR_FLOAT = "MAPPING_STR_FLOAT"
+    BYTES = "BYTES"
+
+
+class Record(ABC):
+    @property
+    def data(self) -> record_data:
+        pass
+
+    @property
+    def record_type(self) -> RecordType:
+        pass
+
+
+class BlobRecord(Record):
+    def __init__(self, data: blob_record_data) -> None:
+        self._record_type = RecordType.BYTES
+
+        try:
+            self._data = self._validate_data(data)
```

Comment:
I might be a naive guy here but what's the difference asking here `if instance(data, blob_record_data)` instead of validate data?

Reply:
It's a good question, there's probably no practical difference—_except_ for the fact that the call to `self._validate_data` uses the typing alias as basis for validation. So if we did

`blob_record_data = Union[bytes, bytearray]` the proposed validation would not require changes. But you cannot combine type aliases and `isinstance`, so `isinstance(data, Union[bytes, bytearray])` does not work, and thus you'd need to change validation as well to `isinstance(data, (bytes, bytearray))`.

Finally, this is ca what pydantic did, which I have replaced.

Reply:
Thanks! You've convinced me there :)